wolfssl/wolfcrypt/types.h

```diff
@@ -643,6 +643,8 @@ decouple library dependencies with standard string, memory and so on.
                 #define XSTRNCASECMP(s1,s2,n) strnicmp((s1),(s2),(n))
             #elif defined(WOLFSSL_CMSIS_RTOSv2)
                 #define XSTRNCASECMP(s1,s2,n) strncmp((s1),(s2),(n))
+            #elif !defined(WOLF_C99) && defined(__STDC__)
+                #define XSTRNCASECMP(s1,s2,n) strncmp((s1),(s2),(n))
```

Comment:
Is this really a safe change?

Reply:
I think so. Let me know if you think it's unsafe.

Reply:
My worry with a binding where `XSTRNCASECMP()` is a misnomer, is that a dev might be using `strncasecmp()` to do a maximally inclusive check, to confirm that a disallowed value is not present, or that a required value is present.

E.g., there are some x509 allowed-key-usage labels (particularly the restrictive ones, like `encipherOnly` and ` decipherOnly`) that are checked case-insensitively, but with this change, we would be testing case-sensitively when `-DWOLF_C89`.  Then we'll incorrectly allow usages that should be disallowed, and incorrectly disallow usages that should be allowed.  And that might be exploitable.

Examples:
```
wolfcrypt/src/asn.c:26147:        if (!XSTRNCASECMP(token, "digitalSignature", len))
wolfcrypt/src/asn.c:26149:        else if (!XSTRNCASECMP(token, "nonRepudiation", len) ||
wolfcrypt/src/asn.c:26150:                 !XSTRNCASECMP(token, "contentCommitment", len))
wolfcrypt/src/asn.c:26152:        else if (!XSTRNCASECMP(token, "keyEncipherment", len))
wolfcrypt/src/asn.c:26154:        else if (!XSTRNCASECMP(token, "dataEncipherment", len))
wolfcrypt/src/asn.c:26156:        else if (!XSTRNCASECMP(token, "keyAgreement", len))
wolfcrypt/src/asn.c:26158:        else if (!XSTRNCASECMP(token, "keyCertSign", len))
wolfcrypt/src/asn.c:26160:        else if (!XSTRNCASECMP(token, "cRLSign", len))
wolfcrypt/src/asn.c:26162:        else if (!XSTRNCASECMP(token, "encipherOnly", len))
wolfcrypt/src/asn.c:26164:        else if (!XSTRNCASECMP(token, "decipherOnly", len))
wolfcrypt/src/asn.c:26207:        if (!XSTRNCASECMP(token, "any", len))
wolfcrypt/src/asn.c:26209:        else if (!XSTRNCASECMP(token, "serverAuth", len))
wolfcrypt/src/asn.c:26211:        else if (!XSTRNCASECMP(token, "clientAuth", len))
wolfcrypt/src/asn.c:26213:        else if (!XSTRNCASECMP(token, "codeSigning", len))
wolfcrypt/src/asn.c:26215:        else if (!XSTRNCASECMP(token, "emailProtection", len))
wolfcrypt/src/asn.c:26217:        else if (!XSTRNCASECMP(token, "timeStamping", len))
wolfcrypt/src/asn.c:26219:        else if (!XSTRNCASECMP(token, "OCSPSigning", len))
wolfcrypt/src/ecc.c:3855:                XSTRNCASECMP(ecc_sets[curve_idx].name, curveName, len) == 0) {
wolfcrypt/src/evp.c:4170:        if (XSTRNCASECMP(name, al->alias, XSTRLEN(al->alias)+1) == 0) {
wolfcrypt/src/evp.c:4178:        if (XSTRNCASECMP(name, ent->name, XSTRLEN(ent->name)+1) == 0) {
```

Reply:
Note that, on close inspection, `wc_SetKeyUsage()` is _already_ buggy, in that length of the `token`s in the supplied arg (`value`) are never checked to make sure they match those of the search strings.  `strncmp()` and friends stop their string walk before the terminating null(s) (that's the whole point of them), so without an additional check, "key" will match "keyEncipherment", etc.

@dgarske we need to fix that and scour the codebase for other instances of this buggy code pattern.  A while back I fixed a batch of these elsewhere, but apparently there's more.